pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,413 @@
+	logger             log.FieldLogger
+	eventRecorder      events.Recorder
+	infrastructureName string
+	region             string
+}
+
+var _ reconcile.Reconciler = &s3EndpointReconciler{}
+
+func (r *s3EndpointReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.logger.Info("reconciling AWS S3 OIDC discovery endpoint on detected change to service account token signer")
+
+	skip, err := r.reconcileClusterAuthentication()
+	if err != nil {
+		r.logger.Error(err)
+		return reconcile.Result{Requeue: true}, err
+	}
+	if skip {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.reconcileS3Endpoint()
+	if err != nil {
+		r.logger.Error(err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *s3EndpointReconciler) reconcileClusterAuthentication() (bool, error) {
+	r.logger.Debugf("reconciling cluster Authentication serviceAccountIssuer")
+
+	auth, err := r.configclientset.ConfigV1().Authentications().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	serviceAccountIssuer := r.getIssuerURL()
+	if auth.Spec.ServiceAccountIssuer != "" {
+		r.logger.Debug("Authentication ServiceAccountIssuer already specified")
+		if auth.Spec.ServiceAccountIssuer != serviceAccountIssuer {
+			r.logger.Debug("Authentication ServiceAccountIssuer is not set to the s3 bucket location, skipping OIDC reconciliation")
+			return true, nil
+		}
+		return false, nil
+	}
+
+	auth.Spec.ServiceAccountIssuer = serviceAccountIssuer
+	_, err = r.configclientset.ConfigV1().Authentications().Update(context.TODO(), auth, metav1.UpdateOptions{})
+	return false, err
+}
+
+func (r *s3EndpointReconciler) getIssuerURL() string {
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s", r.region, r.getBucketName())
+}
+
+func (r *s3EndpointReconciler) getBucketName() string {
+	return fmt.Sprintf("%s-oidc", r.infrastructureName)
+}
+
+func (r *s3EndpointReconciler) reconcileS3Endpoint() error {
+	r.logger.Debugf("reconciling S3 endpoint resources")
+
+	// Get the root secret and create an AWS client
+	secret, err := r.kubeclientset.CoreV1().Secrets(constants.CloudCredSecretNamespace).Get(context.TODO(), awsannotator.AWSCloudCredSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	accessKey, ok := secret.Data[awsannotator.AwsAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing AWS_ACCESS_KEY_ID from cloud cred secret")
+	}
+
+	secretKey, ok := secret.Data[awsannotator.AwsSecretAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing AWS_ACCESS_KEY_ID from cloud cred secret")
+	}
+
+	awsClient, err := aws.NewClient(accessKey, secretKey, r.region, r.infrastructureName)
+	if err != nil {
+		return fmt.Errorf("error creating aws client: %v", err)
+	}
+
+	// Ensure bucket exists
+	bucketName := r.getBucketName()
+	createBucketInput := &s3.CreateBucketInput{
+		Bucket: awssdk.String(bucketName),
```

Comment:
Should this bucket be tagged with the infra ID for cleanup on deprovision?

Reply:
ah, i was wondering about that was done :+1: